hal/src/i2c.rs

```diff
@@ -0,0 +1,412 @@
+                    }
+                }
+
+                pub fn new(i2c: $I2CX, mut pins: (SCL, SDA), freq: Hertz, rcc: &mut RCC, pullup: bool) -> Self
+                    where
+                    SCL: crate::gpio::sealed::$I2cXScl,
+                    SDA: crate::gpio::sealed::$I2cXSda,
+                    {
+                        assert!(freq.integer() <= 1_000_000);
+
+                        Self::enable_clock(rcc);
+                        pins.0.$i2cXsclAf(pullup);
+                        pins.1.$i2cXsdaAf(pullup);
+
+                        // TODO review compliance with the timing requirements of I2C
+                        // t_I2CCLK = 1 / PCLK1
+                        // t_PRESC  = (PRESC + 1) * t_I2CCLK
+                        // t_SCLL   = (SCLL + 1) * t_PRESC
+                        // t_SCLH   = (SCLH + 1) * t_PRESC
+                        //
+                        // t_SYNC1 + t_SYNC2 > 4 * t_I2CCLK
+                        // t_SCL ~= t_SYNC1 + t_SYNC2 + t_SCLL + t_SCLH
+                        let i2cclk = Self::clock(rcc).0;
+                        let ratio = i2cclk / freq.integer() - 4;
+                        let (presc, scll, sclh, sdadel, scldel) = if freq >= 100.kHz() {
+                            // fast-mode or fast-mode plus
+                            // here we pick SCLL + 1 = 2 * (SCLH + 1)
+                            let presc = ratio / 387;
+
+                            let sclh = ((ratio / (presc + 1)) - 3) / 3;
+                            let scll = 2 * (sclh + 1) - 1;
+
+                            let (sdadel, scldel) = if freq > 400.kHz() {
+                                // fast-mode plus
+                                let sdadel = 0;
+                                let scldel = i2cclk / 4_000_000 / (presc + 1) - 1;
+
+                                (sdadel, scldel)
+                            } else {
+                                // fast-mode
+                                let sdadel = i2cclk / 8_000_000 / (presc + 1);
+                                let scldel = i2cclk / 2_000_000 / (presc + 1) - 1;
+
+                                (sdadel, scldel)
+                            };
+
+                            (presc, scll, sclh, sdadel, scldel)
+                        } else {
+                            // standard-mode
+                            // here we pick SCLL = SCLH
+                            let presc = ratio / 514;
+
+                            let sclh = ((ratio / (presc + 1)) - 2) / 2;
+                            let scll = sclh;
+
+                            let sdadel = i2cclk / 2_000_000 / (presc + 1);
+                            let scldel = i2cclk / 800_000 / (presc + 1) - 1;
+
+                            (presc, scll, sclh, sdadel, scldel)
+                        };
+
+                        assert!(presc < 16);
+                        assert!(scldel < 16);
+                        assert!(sdadel < 16);
```

Comment:
I need to read more about the clocks; are these user-accessible asserts or internal-correctness asserts?

If the user can hit these add them to the documentation for panic conditions.

Reply:
They are for internal correctness, as those registers are only 4 bits wide. I will add a note in the documentation that it will panic if the resulting timing registers are out of range, which includes also the next 2 unwraps.